ARCH components. (#951)
ietf-tapswg/api-drafts: draft-ietf-taps-arch.md

+-----------------------------------------------------+
                          |
+-----------------------------------------------------+

> |         Transport  Services Implementation          |

Reviewer: ```suggestion
|          Transport Services Implementation          |
```

## Select Equivalent Protocol Stacks {#equivalence}

A Transport System implementation can select Protocol Stacks based on the Selection and Connection Properties communicated by the application, along with any security parameters. If two different Protocol Stacks can be safely swapped, or raced in parallel (see {{racing}}), then they are considered to be "equivalent". Equivalent Protocol Stacks are defined as stacks that can provide the same Transport Properties and interface expectations as requested by the application.

The following two examples show non-equivalent Protocol Stacks:

>  - If the application requires preservation of message boundaries, a Protocol Stack that runs UDP as the top-level interface to the application is not equivalent to a Protocol Stack that runs TCP as the top-level interface. A UDP stack would allow an application to read out message boundaries based on datagrams sent from the Remote Endpoint, whereas TCP does not preserve message boundaries on its own, but needs a framing protocol on top to determine message boundaries.

Author: sure, done.

The following example shows equivalent Protocol Stacks:

- If the application does not require reliable transmission of data, then a Protocol Stack that adds reliability could be regarded as an equivalent Protocol Stack as long as providing this would not conflict with any other application-requested properties.

To ensure that security protocols are not incorrectly swapped, a Transport System implemenation MUST only select Protocol Stacks that meet application requirements ({{?RFC8922}}). Transport System implemenation SHOULD only race Protocol Stacks where the transport security protocols within the stacks are identical. A Transport System implemenation MUST NOT automatically fall back from secure protocols to insecure protocols, or to weaker versions of secure protocols. A Transport System implementation MAY allow applications to explicitly specify that fallback to a specific other version of a protocol is allowed, e.g., to allow fallback to TLS 1.2 if TLS 1.3 is not available.

## Maintain Interoperability

> It is important to note that neither the Transport Services API {{!I-D.ietf-taps-interface}} nor the guidelines for implementation of the Transport System {{?I-D.ietf-taps-impl}} define new protocols or protocol capabilities that affect what is communicated across the network. Use of the Transport Service Architecture MUST NOT require that a peer on the other side of a connection uses the same API or implementation. A Transport System acting as a connection initiator is able to communicate with any existing ednpoint that implements the transport protocol(s) and all the required properties selected by the Transport System. Similarly, a Transport System acting as a listener can receive connections for any protocol that is supported by the Transport System from an existing initiator that implements the protocol, independent of whether the initiator uses the Transport Services Archhitecture or not.

Reviewer 2: ```suggestion
It is important to note that neither the Transport Services API {{!I-D.ietf-taps-interface}} nor the guidelines for implementation of the Transport System {{?I-D.ietf-taps-impl}} define new protocols or protocol capabilities that affect what is communicated across the network. Use of the Transport Services Architecture MUST NOT require that a peer on the other side of a connection uses the same API or implementation. A Transport System acting as a connection initiator is able to communicate with any existing endpoint that implements the transport protocol(s) and all the required properties selected by the Transport System. Similarly, a Transport System acting as a listener can receive connections for any protocol that is supported by the Transport System from an existing initiator that implements the protocol, independent of whether the initiator uses the Transport Services Architecture or not.
```

> The Transport System makes decisions that select protocols and interfaces. In normal use, a Transport Ssystem SHOULD result in consistent protocol and interface selection decisions for the same network conditions given the same set of Properties. This is intended to provide predictable outcomes to the application using the API.

Reviewer 2: ```suggestion
The Transport System makes decisions that select protocols and interfaces. In normal use, a Transport System SHOULD result in consistent protocol and interface selection decisions for the same network conditions given the same set of Properties. This is intended to provide predictable outcomes to the application using the API.
```

The architecture divides the concepts for Transport Services into two categories:

1. API concepts, which are intended to be exposed to applications; and

> 2. Implementation concepts that are intended to be internally used when building Transport System that implements the Transport Services Architecture.

Reviewer 2: ```suggestion
2. Implementation concepts that are intended to be internally used when building a Transport System that implements the Transport Services Architecture.
```

* Transport Properties: Transport Properties allow the application to express their requirements, prohibitions, and preferences and configure the Transport System.  There are three kinds of Transport Properties: 

  * Selection Properties ({{preestablishment}}): Selection Properties can only be specified on a Preconnection.

  * Connection Properties ({{preestablishment}}): Connection Properties can be specified on a Preconnection and changed on the Connection.

  * Message Properties ({{datatransfer}}): Message Properties can be specified as defaults on a Preconnection or a Connection, and can also be specified during data transfer to affect specific Messages.

> * Connection: A Connection object represents one or more active transport protocol instances that can send and/or receive Messages between local and remote ednpoints. It is an abstraction that represents the communication. The Connection object holds state pertaining to the underlying transport protocol instances and any ongoing data transfers. For example, an active Connection can represent a connection-oriented protocol such as TCP, or can represent a fully-specified 5-tuple for a connectionless protocol such as UDP, where the Connection remains an abstraction at the end points. It can also represent a pool of transport protocol instances, e.g., a set of TCP and QUIC connections to equivalent Endpoints, or a stream of a multi-streaming transport protocol instance. Connections can be created from a Preconnection or by a Listener. 

Reviewer 2: ```suggestion
* Connection: A Connection object represents one or more active transport protocol instances that can send and/or receive Messages between local and remote endpoints. It is an abstraction that represents the communication. The Connection object holds state pertaining to the underlying transport protocol instances and any ongoing data transfers. For example, an active Connection can represent a connection-oriented protocol such as TCP, or can represent a fully-specified 5-tuple for a connectionless protocol such as UDP, where the Connection remains an abstraction at the end points. It can also represent a pool of transport protocol instances, e.g., a set of TCP and QUIC connections to equivalent endpoints, or a stream of a multi-streaming transport protocol instance. Connections can be created from a Preconnection or by a Listener. 
```

You haven't capitalized "endpoints" elsewhere, and I think it's better without capitals (unless it's a defined term in the arch document? I guess it is... hmmm ... then you have to use it with capitals but only after its definition?  oh my oh my...)

* Security Parameters: Security Parameters define an application's requirements for authentication and encryption on a Connection. They are used by Transport Security protocols (such as those described in {{?RFC8922}}) to establish secure Connections. Examples of parameters that can be set include local identities, private keys, supported cryptographic algorithms, and requirements for validating trust of remote identities. Security Parameters are primarily associated with a Preconnection object, but properties related to identities can be associated directly with Endpoints.

### Establishment Actions {#establishment}

* Initiate: The primary action that an application can take to create a Connection to a Remote Endpoint, and prepare any required local or remote state to enable the transmission of Messages. For some protocols, this will initiate a client-to-server style handshake; for other protocols, this will just establish local state (e.g., with connectionless protocols such as UDP). The process of identifying options for connecting, such as resolution of the Remote Endpoint, occurs in response to the Initiate call.

> * Listen: Enables a listener to accept incoming Connections. The Listener will then create Connection objects as incoming connections are accepted ({{events}}). Listeners by default register with multiple paths, protocols, and Local Endpoints, unless constrained by Selection Properties and/or the specified Local Endpoint(s). Connections can be accepted on any of the available paths or Endpoints.

Reviewer 2: ```suggestion
* Listen: Enables a listener to accept incoming Connections. The Listener will then create Connection objects as incoming connections are accepted ({{events}}). Listeners by default register with multiple paths, protocols, and Local Endpoints, unless constrained by Selection Properties and/or the specified Local Endpoint(s). Connections can be accepted on any of the available paths or endpoints.
```

See above for a discussion of "endpoints" vs "Endpoints".

* Connection Properties: The Connection Properties are used to configure protocol-specific options and control per-connection behavior of the Transport System; for example, a protocol-specific Connection Property can express that if TCP is used, the implementation ought to use the User Timeout Option. Note that the presence of such a property does not require that a specific protocol will be used. In general, these properties do not explicitly determine the selection of paths or protocols, but can be used by an implementation during connection establishment. Connection Properties are specified on a Preconnection prior to Connection establishment, and can be modified on the Connection later. Changes made to Connection Properties after Connection establishment take effect on a best-effort basis.

> * Security Parameters: Security Parameters define an application's requirements for authentication and encryption on a Connection. They are used by Transport Security protocols (such as those described in {{?RFC8922}}) to establish secure Connections. Examples of parameters that can be set include local identities, private keys, supported cryptographic algorithms, and requirements for validating trust of remote identities. Security Parameters are primarily associated with a Preconnection object, but properties related to identities can be associated directly with Endpoints.

Reviewer 2: Sorry, not this PR's problem, but I just noticed a capitalized Endpoint here:

```suggestion
* Security Parameters: Security Parameters define an application's requirements for authentication and encryption on a Connection. They are used by Transport Security protocols (such as those described in {{?RFC8922}}) to establish secure Connections. Examples of parameters that can be set include local identities, private keys, supported cryptographic algorithms, and requirements for validating trust of remote identities. Security Parameters are primarily associated with a Preconnection object, but properties related to identities can be associated directly with endpoints.
```

* Message Properties: Message Properties are used to specify details about Message transmission. They can be specified directly on individual Messages, or can be set on a Preconnection or Connection as defaults. These properties might only apply to how a Message is sent (such as how the transport will treat prioritization and reliability), but can also include properties that specific protocols encode and communicate to the Remote Endpoint. When receiving Messages, Message Properties can contain information about the received Message, such as metadata generated at the receiver and information signalled by the Remote Endpoint. For example, a Message can be marked with a Message Property indicating that it is the final message on a connection if the peer sent a TCP FIN.

* Send: The action to transmit a Message over a Connection to the Remote Endpoint. The interface to Send can accept Message Properties specific to how the Message content is to be sent. The status of the Send operation is delivered back to the sending application in an Event ({{events}}).

> * Receive: An action that indicates that the application is ready to asynchronously accept a Message over a Connection from a remote endpoint, while the Message content itself will be delivered in an Event ({{events}}). The interface to Receive can include Message Properties specific to the Message that is to be delivered to the application.

Reviewer 2: ```suggestion
* Receive: An action that indicates that the application is ready to asynchronously accept a Message over a Connection from a Remote Endpoint, while the Message content itself will be delivered in an Event ({{events}}). The interface to Receive can include Message Properties specific to the Message that is to be delivered to the application.
```

So "Remote Endpoint" is capitalized but "endpoint" is not? At least for now, that's how it seems to be...

When the API clones an existing Connection, this adds a new Connection to the Connection Group. A change to one of the Connection Properties on any Connection in the Connection Group automatically changes the Connection Property for all others. All Connections in a Connection Group share the same set of Connection Properties except for the Connection Priority. These Connection Properties are said to be entangled. 

For multiplexing transport protocols, only Connections within the same Connection Group are allowed to be multiplexed together. Passive Connections can also be added to a Connection Group, e.g., when a Listener receives a new Connection that is just a new stream of an already active multi-streaming protocol
instance.

While Connection Groups are managed by the Transport System, an application can define Connection Contexts to control caching boundaries, as discussed in {{conn-context}}.

## Transport System Implementation

> This section defines the key concepts of the Transport System component of the Transport Services Architecture.

Reviewer 2: This gets odd. If the Transport System is not a complete instantiation of the architecture (as I thought up to here), but just a component, then how can it be one of the "key concepts" of itself?  because it's also the next item in the list.

Author: Made the Transport System more clearly sit below the API when first introduced earlier (subtle because the "API" is only realised when you have a system designed to implement it, but at least tries to make reading less painful. We didn't talk about a "component" here, so suggest we don't. Text updated.

When the API clones an existing Connection, this adds a new Connection to the Connection Group. A change to one of the Connection Properties on any Connection in the Connection Group automatically changes the Connection Property for all others. All Connections in a Connection Group share the same set of Connection Properties except for the Connection Priority. These Connection Properties are said to be entangled. 

For multiplexing transport protocols, only Connections within the same Connection Group are allowed to be multiplexed together. Passive Connections can also be added to a Connection Group, e.g., when a Listener receives a new Connection that is just a new stream of an already active multi-streaming protocol
instance.

While Connection Groups are managed by the Transport System, an application can define Connection Contexts to control caching boundaries, as discussed in {{conn-context}}.

## Transport System Implementation

> This section defines the key concepts of the Transport System of the Transport Services Architecture.

Reviewer 2: I still don't like the recursion, where the Transport System is a "key concept" of the Transport System. Are these maybe just key concepts of the Transport Services Architecture?

and applications ought not to use private keys intended for server
authentication as keys for client authentication.

A Transport System implementing the Transport Services Architecture must not automatically fall back from
secure protocols to insecure protocols, or to weaker versions of secure
protocols (see {{equivalence}}). For example, if an application requests a specific version of TLS,
but the desired version of TLS is not available, its connection will fail.
Applications are thus responsible for implementing security protocol fallback
or version fallback by creating multiple Connections, if so
desired. Alternatively, the Transport Services API MAY allow applications to

> specify that the Transport System is allowed to fallback to a specific other version of a protocol.

Reviewer 2: in the replaced text, "fallback" is used as a noun, but here it is a verb - I'm not sure it's a verb that exists? Maybe "falling back" ?
